### PR TITLE
mdm clear move to batch2

### PR DESCRIPTION
**Pull request by fil512**

### CODE REVIEW

File: hapi-fhir-storage/src/main/java/ca/uhn/fhir/jpa/api/pid/HomogeneousResourcePidList.java [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+	final String myResourceType;
+
+	public HomogeneousResourcePidList(String theResourceType, Collection<ResourcePersistentId> theIds, Date theLastDate) {
+		super(theIds, theLastDate);
+		myResourceType = theResourceType;
+	}
+
+	@Override
+	public String getResourceType(int i) {
+			return myResourceType;
```
Comment (TipzCM) [inline hapi-fhir-storage/src/main/java/ca/uhn/fhir/jpa/api/pid/HomogeneousResourcePidList.java:23]: return getResourceType(); ? (nitpick)

**Review by TipzCM — CHANGES_REQUESTED** [review]
minor stuff, but some of it i think should be fixed first

File: hapi-fhir-jpaserver-mdm/src/test/java/ca/uhn/fhir/jpa/mdm/provider/MdmProviderClearLinkR4Test.java [hunk 2]
```diff
@@ -74,7 +75,8 @@ public void testClearPatientLinks() {
 		try {
 			myPatientDao.read(new IdDt(mySourcePatientId.getValueAsString()).toVersionless());
 			fail();
-		} catch (ResourceNotFoundException e) {
+		} catch (ResourceGoneException e) {
+			// Expected exception
```
[the remark this answers is not in the document]
Comment (fil512) [reply hapi-fhir-jpaserver-mdm/src/test/java/ca/uhn/fhir/jpa/mdm/provider/MdmProviderClearLinkR4Test.java:79]: the assertion is that the patient is now gone.  the assertion's in two parts, the `fail()` and the ``catch`

File: hapi-fhir-server-mdm/src/main/java/ca/uhn/fhir/mdm/api/IMdmLinkSvc.java [hunk 3]
```diff
@@ -45,4 +48,6 @@ public interface IMdmLinkSvc {
 	 * @param theMdmTransactionContext
 	 */
 	void deleteLink(IAnyResource theExistingGoldenResource, IAnyResource theSourceResource, MdmTransactionContext theMdmTransactionContext);
+
+	void deleteLinksWithAnyReferenceTo(List<ResourcePersistentId> thePersistentIds);
```
[the remark this answers is not in the document]
Comment (fil512) [reply hapi-fhir-server-mdm/src/main/java/ca/uhn/fhir/mdm/api/IMdmLinkSvc.java:52]: added javadoc.  you're right all interface methods should include javadoc

File: hapi-fhir-storage-batch2/src/main/java/ca/uhn/fhir/batch2/config/Batch2Constants.java [hunk 4]
```diff
@@ -0,0 +1,9 @@
+package ca.uhn.fhir.batch2.config;
+
+import org.hl7.fhir.r4.model.InstantType;
+
+import java.util.Date;
+
+public class Batch2Constants {
+	public static final Date BATCH_START_DATE = new InstantType("2000-01-01T00:00:00Z").getValue();
```
[the remark this answers is not in the document]
Comment (fil512) [reply hapi-fhir-storage-batch2/src/main/java/ca/uhn/fhir/batch2/config/Batch2Constants.java:8]: This date was hardcoded in the system before.  I pulled it out into a constant.  Added javadoc.

File: hapi-fhir-storage-batch2/src/main/java/ca/uhn/fhir/batch2/jobs/chunk/ResourceIdListWorkChunkJson.java [hunk 5]
```diff
@@ -0,0 +1,83 @@
...
+		}
+
+		return myTypedPids
+			.stream()
+			.map(t -> new ResourcePersistentId(t.getPid()))
+			.collect(Collectors.toList());
+	}
+
+	public int size() {
+		return myTypedPids.size();
```
[the remark this answers is not in the document]
Comment (fil512) [reply hapi-fhir-storage-batch2/src/main/java/ca/uhn/fhir/batch2/jobs/chunk/ResourceIdListWorkChunkJson.java:72]: Good call!

File: hapi-fhir-storage-batch2/src/main/java/ca/uhn/fhir/batch2/jobs/step/IIdChunkProducer.java [hunk 6]
```diff
@@ -0,0 +1,13 @@
...
+import ca.uhn.fhir.batch2.jobs.chunk.ChunkRangeJson;
+import ca.uhn.fhir.interceptor.model.RequestPartitionId;
+import ca.uhn.fhir.jpa.api.pid.IResourcePidList;
+
+import javax.annotation.Nonnull;
+import javax.annotation.Nullable;
+import java.util.Date;
+
+public interface IIdChunkProducer<IT extends ChunkRangeJson> {
+	IResourcePidList fetchResourceIdsPage(Date theNextStart, Date theEnd, @Nonnull Integer thePageSize, @Nullable RequestPartitionId theRequestPartitionId, IT theData);
```
[the remark this answers is not in the document]
Comment (fil512) [reply hapi-fhir-storage-batch2/src/main/java/ca/uhn/fhir/batch2/jobs/step/IIdChunkProducer.java:12]: Excellent questions.  I have added plenty of javadoc.

File: hapi-fhir-storage-mdm/src/main/java/ca/uhn/fhir/mdm/batch2/MdmBatch2Config.java [hunk 7]
```diff
@@ -0,0 +1,29 @@
...
+	@Autowired
+	JobDefinitionRegistry myJobDefinitionRegistry;
+
+	@Autowired
+	ApplicationContext myApplicationContext;
+
+	@PostConstruct
+	public void start() {
+		JobDefinition jobDefinition = myApplicationContext.getBean(MDM_CLEAR_JOB_BEAN_NAME, JobDefinition.class);
+		myJobDefinitionRegistry.addJobDefinition(jobDefinition);
```
[the remark this answers is not in the document]
Comment (fil512) [reply hapi-fhir-storage-mdm/src/main/java/ca/uhn/fhir/mdm/batch2/MdmBatch2Config.java:27]: if logs a warning message and skips the registration if it's already registered.

File: hapi-fhir-storage/src/main/java/ca/uhn/fhir/jpa/api/pid/HomogeneousResourcePidList.java (hunk 1 above)
Comment (fil512) [reply hapi-fhir-storage/src/main/java/ca/uhn/fhir/jpa/api/pid/HomogeneousResourcePidList.java:23]: I'm not accustomed to coding this way, but it does feel more polymorphic with the accessor.  Changed.

**Review by tadgh — APPROVED** [review]
Lgtm

**Pull request merged**